packages/kit/src/runtime/app/forms.js

```diff
@@ -0,0 +1,85 @@
+import { goto, invalidateAll } from './navigation.js';
+import { client } from '../client/singletons.js';
+
+/**
+ * @param {string} name
+ */
+function guard(name) {
+	return () => {
+		throw new Error(`Cannot call ${name}(...) on the server`);
+	};
+}
+
+const ssr = import.meta.env.SSR;
+
+/** @type {import('$app/forms').applyAction} */
+export const applyAction = ssr ? guard('applyAction') : client.apply_action;
+
+/** @type {import('$app/forms').enhance} */
+export function enhance(form, submit = () => {}) {
+	/** @param {import('$app/forms').SubmissionResult} result */
+	const fallback_callback = (result) => {
+		if (
+			(result.type === 'success' || result.type === 'invalid') &&
+			location.origin + location.pathname === form.action.split('?')[0]
+		) {
+			applyAction(result);
+		}
+
+		if (result.type === 'success') {
+			invalidateAll();
+		} else if (result.type === 'redirect') {
+			goto(result.location);
+		} else if (result.type === 'error') {
+			console.error(result.error);
+		}
+	};
+
+	/** @type {unknown} */
+	let current_token;
```

Comment:
this feels like something that should be left to userland — we shouldn't be preventing multiple submissions as a blanket policy